### PR TITLE
feat(fs.lua): pass path to vim.fs.find function predicate and lazy evaluate the predicate

### PR DESCRIPTION
**Pull request by mike325**
Predicates can do more complex filters if they know were the file/directory is located, right now there's no easy way to get files under certain directories (ex: grab all files under `test/`) or exclude the content of certain paths (ex. `build/`, `.git/`)

### CODE REVIEW

File: runtime/lua/vim/fs.lua
Comment (gpanders) [inline runtime/lua/vim/fs.lua:147]: Can you add a comment in the docstring about what `{path}` is. Maybe something like

```
If {names} is a function, it is called for each file and directory within the traversed directory and is passed the base name of each file or directory and the path of the directory being traversed. The function should return `true` if the given file or directory is considered a match.
```

**Comment by Mike325** [comment]
I see `news.txt` check is failing, should I include the new parameter somewhere or just ignore the failed check?

**Comment by clason** [comment]
It's mainly meant as a reminder, it doesn't need to pass for the PR to be merged.

In this case, I defer to @gpanders whether the change is notable enough to warrant an entry.

**Comment by gpanders** [comment]
I don't think a news entry is necessary in this case.

File: runtime/doc/lua.txt
Comment (gpanders) [inline runtime/doc/lua.txt:2365]: Sorry if I wasn't clear, I was suggesting that this should be in the description of the `{names}` parameter below.

My suggestion was a bit prolix, feel free to condense it. Really all I want to see is that someone who's never used this function before can read the docstring and have clarity on what it means for `{names}` to be a function and how that function works (i.e. it takes two arguments, what those arguments are, and what the function should return).

This is also a case where an example is worth a thousand words.

File: runtime/doc/lua.txt
Comment (clason) [inline runtime/doc/lua.txt:2365]: I think this example should include the `path` key since that is necessary for getting the correct LSP root directory (the main motivation for this example , and in fact this function).

Here's the function I use in my config (`file` is `args.file` from a `FileType` autocommand):
```lua
local find_root = function(markers, file)
  return vim.fs.dirname(vim.fs.find(markers, { upward = true, path = vim.fn.fnamemodify(file, ':p:h') })[1])
end
```

File: runtime/lua/vim/fs.lua
Comment (gpanders) [inline runtime/lua/vim/fs.lua:171]: Nit:
```suggestion
---             If {names} is a function, it will be called once for each file or directory
```

**Review by gpanders — APPROVED** [review]
Thanks!

**Comment by gpanders** [comment]
@Mike325 Could you also add a test case for this? You can add a simple check here: https://github.com/neovim/neovim/blob/f64098a2df774c79dd454f63ac491570cdcaf2b2/test/functional/lua/fs_spec.lua#L240-L255

File: runtime/doc/lua.txt
Comment (dundargoc) [inline runtime/doc/lua.txt:2377]: Nitpick: this assumes the reader knows what c++ files look like. Something like "get all files ending in .cpp and .hpp" or "get all files with extension .cpp and .hpp" would be preferable IMO

**Review by gpanders — APPROVED** [review]

**Review by justinmk — APPROVED** [review]

File: runtime/lua/vim/fs.lua
Comment (justinmk) [inline runtime/lua/vim/fs.lua:173]: The wording is confusing.

```suggestion
---             If {names} is a function, it is called for each traversed file or directory
---             with args:
---             - name: base name of the current item
---             - path: full path of the current item
```

File: runtime/lua/vim/fs.lua
Comment (justinmk) [inline runtime/lua/vim/fs.lua:174]: I guess a blank line is needed here to work around the docs rendering issue (it joins this line).
```suggestion
---
---             The function should return `true` if the given file or directory is considered a match.
```

Comment (Mike325) [reply runtime/lua/vim/fs.lua:174]: This does not seem to work, `gen_vimdoc.py` sends everything after the new line to the general description, just below the code examples, it renders like 

```help
...
     -- get all files ending with .cpp or .hpp inside lib/
     local cpp_hpp = vim.fs.find(function(name, path)
       return name:match('.*%.[ch]pp$') and path:match('[/\\]lib$')
     end, {limit = math.huge, type = 'file'})
<

    The function should return `true` if the given file or directory is considered a match.

    Parameters: ~
      • {names}  (string|table|fun(name: string, path: string): boolean) Names
                 of the files and directories to find. Must be base names,
                 paths and globs are not supported when {names} is a string or
                 a table. If {names} is a function, it is called for each
                 traversed file and directory with args:
                 • name: base name of the current item
                 • path: full path of the current item
...
```

Comment (justinmk) [reply runtime/lua/vim/fs.lua:174]: ok, I'll try to fix it later. 

Don't want to block this PR meanwhile.

**Pull request merged**
